src/pg_query_summary_truncate.c

```diff
@@ -0,0 +1,673 @@
+}
+
+static void add_truncation(TruncationState *state, enum TruncationAttr attr,
+		Node *node, int32_t length)
+{
+	PossibleTruncation *truncation = palloc(sizeof(PossibleTruncation));
+
+	truncation->attr = attr;
+	truncation->node = node;
+	truncation->depth = state->depth;
+	truncation->length = length;
+
+	state->truncations = lappend(state->truncations, truncation);
+}
+
+static void add_truncation_where_clause(TruncationState *state, Node *node, Node *whereClause)
+{
+	if (whereClause == NULL)
+		return;
+
+	add_truncation(state,
+			TRUNCATION_WHERE_CLAUSE,
+			node,
+			where_clause_len(whereClause));
+}
+
+static bool
+summary_truncation_options(Node *node, TruncationState *state)
+{
+	if (node == NULL)
+		return false;
+
+	switch (nodeTag(node))
+	{
+		case T_RawStmt:
+			// For some reason our statement has been trapped in a RawStmt.
+			// Free them.
+			return summary_truncation_options(castNode(RawStmt, node)->stmt, state);
+
+		case T_SelectStmt:
+			{
+				SelectStmt *stmt = castNode(SelectStmt, node);
+
+				if (stmt->targetList != NULL)
+					add_truncation(state,
+							TRUNCATION_TARGET_LIST,
+							node,
+							select_target_list_len(stmt->targetList));
+
+				add_truncation_where_clause(state, node, stmt->whereClause);
+
+				if (stmt->valuesLists != NULL)
+					add_truncation(state,
+							TRUNCATION_VALUES_LISTS,
+							node,
+							select_values_lists_len(stmt->valuesLists));
+
+				break;
+			}
+
+		case T_UpdateStmt:
+			{
+				UpdateStmt *stmt = castNode(UpdateStmt, node);
+
+				if (stmt->targetList != NULL)
+					add_truncation(state,
+							TRUNCATION_TARGET_LIST,
+							node,
+							update_target_list_len(stmt->targetList));
+
+				add_truncation_where_clause(state, node, stmt->whereClause);
+
+				break;
+			}
+
+		case T_DeleteStmt:
+			{
+				DeleteStmt *stmt = castNode(DeleteStmt, node);
+
+				add_truncation_where_clause(state, node, stmt->whereClause);
+
+				break;
+			}
+
+		case T_CopyStmt:
+			{
+				CopyStmt *stmt = castNode(CopyStmt, node);
+
+				add_truncation_where_clause(state, node, stmt->whereClause);
+
+				break;
+			}
+
+		case T_InsertStmt:
+			{
+				InsertStmt *stmt = castNode(InsertStmt, node);
+
+				if (stmt->cols != NULL)
+					add_truncation(state, TRUNCATION_COLS, node, cols_len(stmt->cols));
+
+				break;
+			}
+
+		case T_IndexStmt:
+			{
+				IndexStmt *stmt = castNode(IndexStmt, node);
+				add_truncation_where_clause(state, node, stmt->whereClause);
+				break;
+			}
+
+		case T_RuleStmt:
+			{
+				RuleStmt *stmt = castNode(RuleStmt, node);
+				add_truncation_where_clause(state, node, stmt->whereClause);
+				break;
+			}
+
+		case T_CommonTableExpr:
+			{
+				CommonTableExpr *stmt = castNode(CommonTableExpr, node);
+
+				if (stmt->ctequery != NULL) {
+					PgQueryDeparseResult result = pg_query_deparse(node);
+
+					if (result.error) {
+						state->error = result.error;
+						return false;
+					}
+
+					int32_t length = strlen(result.query);
```

Comment:
I'd probably inline this in the function call below, to match the other `_len` calls (though this is a different kind of function being called, I don't think putting it in a variable is necessary)